v-next/hardhat/src/internal/core/config-validation.ts

```diff
@@ -110,6 +124,63 @@ export function collectValidationErrorsForUserConfig(
   return validationErrors;
 }
 
+export function validatePaths(
+  paths: ProjectPathsUserConfig,
+): HardhatUserConfigValidationError[] {
+  const validationErrors: HardhatUserConfigValidationError[] = [];
+
+  if (paths.cache !== undefined) {
+    validationErrors.push(...validatePath(paths.cache, "cache"));
+  }
+
+  if (paths.artifacts !== undefined) {
+    validationErrors.push(...validatePath(paths.artifacts, "artifacts"));
+  }
+
+  if (paths.tests !== undefined) {
+    // paths.tests of type TestPathsUserConfig is not validated because it is customizable by the user
+    if (!isObject(paths.tests)) {
+      validationErrors.push(...validatePath(paths.tests, "tests"));
+    }
+  }
+
+  if (paths.sources !== undefined) {
+    if (Array.isArray(paths.sources)) {
+      for (const [index, source] of paths.sources.entries()) {
+        validationErrors.push(...validatePath(source, "sources", index));
+      }
+      // paths.sources of type SourcePathsUserConfig is not validated because it is customizable by the user
+    } else if (!isObject(paths.sources)) {
+      validationErrors.push(...validatePath(paths.sources, "sources"));
+    }
+  }
+
+  return validationErrors;
+}
+
+function validatePath(
+  filePath: unknown,
+  pathName: "cache" | "artifacts" | "tests" | "sources",
+  index?: number, // If the filePath is part of an array, pass its index to make the error message more specific
```

Comment:
```suggestion
  index?: number,
```
I think we can leave the comment out, the usage is pretty clear; and the function is not exported so the invocation in practice is in the function above.

I think you either go with a full JSDOC comment on the `validatePath` function to include the param, or no comment and rely on context. Both work.